alphagov/paas-bootstrap
# Upgrade to terraform 0.12.x 173847170

What
----

Upgrade Terraform to version 0.12.29, keeping the AWS provider at 2.x.
- This was mostly painless - just ran `terraform 0.12 upgrade` across all projects.
- Created version_constraints.tf and providers.tf as global information linked into all projects.
- Adapted some scripts to be able to digest both terraform state file version 3 and 4.
- Preserving AWS provider 2.x due to a [bug in handling domain_validation_options in AWS provider 3.x](https://github.com/terraform-providers/terraform-provider-aws/issues/14447).
- From version 0.12, `terraform validate` checks that files included by the `file` function do exist at validation time. Added an empty `terraform/bosh/id_rsa.pub` to make validation pass.

Not keeping up with recent Terraform versions creates tech debt.

How to review
-------------

- Code review
- Create and destroy a bootstrap environment in dev (as there is no SSO required here, you can use a temporary DEPLOY_ENV)
  - `make dev deployer-concourse bootstrap`
  - Start `create-bosh-concourse` pipeline
  - `make dev upload-all-secrets` after the state bucket is created (check that all commands in the script execute successfully, with my Yubikey sometimes SSH keys are not pulled first time).
  - The `bosh-terraform` tasks currently seems to have a race condition for the creation of certain log groups and may need restarting once.
  - Run `destroy-bosh-concourse` pipeline

Who can review
--------------

not @schmie
